docs/faq.rst

```diff
@@ -297,42 +310,48 @@ Do you have tips on creating Dockerfiles?
    client/server issues, it is also not compatible with CWL)
 -  do not depend on changes to ``hostname`` or ``/etc/hosts``, Docker
    will interfere with this
+-  use a well-known and secure :ref:`dict parent image` such as official `debian <https://hub.docker.com/_/debian>`__ or `Python <https://hub.docker.com/_/python>`__ images
 -  try to keep your Docker images small
+   -  however, do not use alpine images, or other images that lack bash, as your parent image unless you will be installing bash in the Dockerfile 
+  
+
+:ref:`(back to top) <topFAQ>`
+
+How should I handle large reference files when designing workflows and Dockerfiles?
+-----------------------------------------------------------------------------------
+Generally speaking, you can choose to either "package" reference files in your Docker image, or you can treat them as "inputs" so they can be stage outside and mounted into the running container. Both of these approaches will work on a variety of backends, but there are some backends in which one option might be better than the other.
+
 
+      - if you are running on a backend that can mount files into your containers quickly, such as a local installation of miniwdl, and your input files are available without having to download them, is probably best to treat reference files as an input
+      - if you are running your workflow on a backend that needs to download input files (such as Terra), and you need to put those reference files into many instances of a Docker container (such as if you are using scattered tasks), it is probably best to bake your reference files into the Dockerfile so they are immediately available in the Docker image
```

Comment:
Hmm, this contradicts the advice  we give on [FAIR best practices](https://docs.dockstore.org/en/stable/advanced-topics/best-practices/best-practices-secure-fair-workflows.html?highlight=fair#image-container-best-practices), where we say "keep images light" and "don't include test data in the image".

I suggest dropping this section entirely, for now, until we sort that out.

You know a ton more about writing workflows than I do, and so this might be the right approach, but if so, then we should update the best FAIR practices as well.

Reply:
## short answer
Yeah, I tried to balance FAIR plus my own experience and it's a bit of a complicated topic. It can be argued that including a reference is more FAIR because it is more reproducible, especially with certain types of tooling. I'm in favor of dropping all references to image size in the short term, if that's what you mean?

## long answer (optional reading, sorry, I'm still verbose)
I came to the conclusion that sometimes it is best to include the reference thanks to working with a tuberculosis toolkit called [clockwork](https://github.com/iqbal-lab-org/clockwork). clockwork has a function ([which invokes this script](https://github.com/iqbal-lab-org/clockwork/blob/master/scripts/download_tb_reference_files.pl)) dedicated to downloading and preparing reference genomes, resulting in the creation of three reference genome folders -- one of which is about 15 GB in size because it contains human, Epstein-Barr, HIV, etc for decontamination purposes plus a ton of metadata. The other tasks expect those reference genome folders to be in the workdir, which is fine when you're working with the toolkit in an interactive way, but if you're trying to make this slightly modular as WDL tasks, you now have to pass around 15 GB of reference files around as inputs. clockwork doesn't have the ability to take in inputs as files, instead expecting them in the workdir, which isn't where WDL executors tend to localize files, so you need to also copy them [from an input folder with potentially odd permissions](https://miniwdl.readthedocs.io/en/latest/runner_advanced.html#read-only-input-files) into the workdir. So, those task-level inputs for the reference files are accessible to users, which means a user who has the reference files lying around could put them in manually... but having those inputs available increases the chance that a user may put the wrong thing in there, since explaining what the reference folder needs to contain can't be done very well in a short docstring. Also, WDL 1.0 does not accept folders as inputs or outputs, so you have to tarball them too, and all subsequent tasks need to include a untar command. [Terra is not very good at getting files in and out of Docker containers quickly](https://support.terra.bio/hc/en-us/community/posts/10732090690843-Switch-Localization-Delocalization-From-gsutil-to-gcloud-storage), so you're wasting compute time and resources on passing that tarball around. If you want to compress the tarball to make things faster, you're going to want to use [pigz](https://zlib.net/pigz/) because gzip is orders of magnitude slower on multi-GB files, but then you have to roll your own Docker image because [clockwork's provided Docker image](https://github.com/iqbal-lab-org/clockwork/pkgs/container/clockwork) doesn't come with pigz. [And if you're rolling your own Docker image, you may as well just include the reference files](https://github.com/aofarrel/clockwork-wdl/blob/main/build_dockers.sh). It certainly is possible to make passing in reference files work -- I did so until [2.9.0 of clockwork-wdl](https://github.com/aofarrel/clockwork-wdl/releases). But I realized including the references was easier for users (less inputs to keep track of) and less likely to break.

...of course, I might be the edge case here. I do know plenty of other tools expect reference files in specific places, so I think there is still a case to be made for including the references in the docker image for other tasks, but I don't want to overextrapolate here. 

(It should also be noted the Terra use case is a bit of a rock and a hard place situation -- Terra is sluggish at passing around 15 GB tarballs, but it doesn't cache Docker images either. That means **every** instance of a scattered task either has to pass in 15 GB of files via a slow gsutil command, or pull a 15 GB Docker image from Docker Hub. I suspect the Docker option is more cost-effective for the user but neither option is particularly good, especially considering Docker Hub sometimes fumbles the `docker pull` which crashes the entire workflow.)

Reply:
> Yeah, I tried to balance FAIR plus my own experience and it's a bit of a complicated topic. It can be argued that including a reference is more FAIR because it is more reproducible, especially with certain types of tooling. I'm in favor of dropping all references to image size in the short term, if that's what you mean?

What I meant is the doc should have the same recommendation everywhere. With this PR, the doc will:

* On the FAIR page, say "don't include data in Docker images", period.
* This page will say "sometimes include data in the Docker images"

I think we should have the same message throughout the doc, and until we do, we shouldn't introduce divergences, even if in the end we end up with your recommendation.

Reply:
Agreed completely and I'll be deleting the "sometimes include data" from this page as brought up for that reason. Thanks for flagging this, I had forgotten that our FAIR page had that recommendation. It's important docs don't contradict.